File: tests/stress/scale-test.ts
/**
 * Wago Scale / Stress Test Skeleton
 *
 * This file outlines the structure for scale and stress tests that will be
 * implemented once the orchestration layer (Phase 3) is operational.
 *
 * Prerequisites:
 *   - Running Wago API server with database
 *   - At least one active WAHA worker provisioned
 *   - Valid test user JWT token
 *   - Environment variables: API_URL, TEST_JWT_TOKEN
 *
 * Run (when implemented):
 *   npx tsx tests/stress/scale-test.ts
 */

// ---------------------------------------------------------------------------
// Configuration
// ---------------------------------------------------------------------------

const CONFIG = {
  apiUrl: process.env.API_URL ?? 'http://localhost:3001',
  testToken: process.env.TEST_JWT_TOKEN ?? '',
  workerCapacityThreshold: 0.8, // 80% triggers scale-up
  workerDrainThreshold: 0.3, // <30% for all workers triggers scale-down
  drainWaitMinutes: 10, // wait time before scale-down
  maxSessionsPerWorker: 50,
};

// ---------------------------------------------------------------------------
// Helpers (to be implemented)
// ---------------------------------------------------------------------------

interface TestConnection {
  id: string;
  sessionName: string;
  workerId: string | null;
  status: string;
}

interface WorkerSnapshot {
  id: string;
  status: string;
  currentSessions: number;
  maxSessions: number;
}

/** Create a connection via the API and return its details. */
async function createConnection(): Promise<TestConnection> {
  // POST /connections with Authorization: Bearer ${CONFIG.testToken}
  // Returns the created connection record
  throw new Error('Not implemented — requires Phase 2 connections controller');
}

/** Delete a connection via the API. */
async function deleteConnection(connectionId: string): Promise<void> {
  // DELETE /connections/:id with Authorization: Bearer ${CONFIG.testToken}
  throw new Error('Not implemented — requires Phase 2 connections controller');
}

/** Get the current list of workers from the internal admin API or DB. */
async function getWorkers(): Promise<WorkerSnapshot[]> {
  // GET /admin/workers (internal endpoint, not user-facing)
  throw new Error('Not implemented — requires Phase 3 orchestration layer');
}

/** Wait for a condition to become true, polling at the given interval. */
async function waitFor(
  description: string,
  conditionFn: () => Promise<boolean>,
  timeoutMs: number = 120_000,
  intervalMs: number = 2_000,
): Promise<void> {
  const start = Date.now();
  while (Date.now() - start < timeoutMs) {
    if (await conditionFn()) return;
    await new Promise((r) => setTimeout(r, intervalMs));
  }
  throw new Error(`Timed out waiting for: ${description}`);
}

/** Simulate sending webhook events at a given rate. */
async function fireWebhookEvents(
  eventsPerSecond: number,
  durationSeconds: number,
): Promise<{ totalSent: number; failures: number; avgLatencyMs: number }> {
  // Pushes synthetic events into the BullMQ webhook queue
  throw new Error('Not implemented — requires Phase 3 event router');
}

// ---------------------------------------------------------------------------
// Test: Scale-Up Trigger
// ---------------------------------------------------------------------------

async function testScaleUpTrigger(): Promise<void> {
  console.log('\n=== Test: Scale-Up Trigger ===');
  console.log(
    `Creating connections until a worker exceeds ${CONFIG.workerCapacityThreshold * 100}% capacity...`,
  );

  const created: TestConnection[] = [];

  try {
    // Step 1: Get initial worker count
    const initialWorkers = await getWorkers();
    const initialWorkerCount = initialWorkers.length;
    console.log(`Initial worker count: ${initialWorkerCount}`);

    // Step 2: Create connections until capacity threshold is hit
    const targetPerWorker = Math.ceil(
      CONFIG.maxSessionsPerWorker * CONFIG.workerCapacityThreshold,
    );

    for (let i = 0; i < targetPerWorker + 5; i++) {
      const conn = await createConnection();
      created.push(conn);
      console.log(
        `  Created connection ${i + 1}/${targetPerWorker + 5}: ${conn.sessionName}`,
      );
    }

    // Step 3: Wait for a new worker to be provisioned
    await waitFor(
      'new worker to be provisioned',
      async () => {
        const workers = await getWorkers();
        return workers.length > initialWorkerCount;
      },
      300_000, // 5 min timeout — provisioning takes time
    );

    const finalWorkers = await getWorkers();
    console.log(
      `Scale-up verified: ${initialWorkerCount} -> ${finalWorkers.length} workers`,
    );
  } finally {
    // Cleanup: remove all test connections
    for (const conn of created) {
      await deleteConnection(conn.id).catch(() => {});
    }
  }
}

// ---------------------------------------------------------------------------
// Test: Scale-Down Trigger
// ---------------------------------------------------------------------------

async function testScaleDownTrigger(): Promise<void> {
  console.log('\n=== Test: Scale-Down Trigger ===');
  console.log(
    'Removing connections until all workers are below ' +
      `${CONFIG.workerDrainThreshold * 100}% capacity, then waiting ` +
      `${CONFIG.drainWaitMinutes} minutes...`,
  );

  // Step 1: Record current workers
  const initialWorkers = await getWorkers();
  if (initialWorkers.length < 2) {
    console.log(
      'SKIP: Need at least 2 workers to test scale-down. ' +
        'Run scale-up test first.',
    );
    return;
  }

  // Step 2: Delete connections until all workers < 30%
  // (Assumes connections can be listed and deleted)

  // Step 3: Wait for drain period + consolidation
  const waitMs = (CONFIG.drainWaitMinutes + 2) * 60 * 1000;
  await waitFor(
    'worker count to decrease after drain period',
    async () => {
      const workers = await getWorkers();
      return workers.filter((w) => w.status === 'active').length <
        initialWorkers.length;
    },
    waitMs,
    15_000,
  );

  const finalWorkers = await getWorkers();
  console.log(
    `Scale-down verified: ${initialWorkers.length} -> ` +
      `${finalWorkers.filter((w) => w.status === 'active').length} active workers`,
  );
}

// ---------------------------------------------------------------------------
// Test: Connection Storm
// ---------------------------------------------------------------------------

async function testConnectionStorm(): Promise<void> {
  console.log('\n=== Test: Connection Storm ===');
  console.log('Creating 100 connections as fast as possible...');

  const STORM_SIZE = 100;
  const startTime = Date.now();

  // Fire all creates in parallel
  const promises = Array.from({ length: STORM_SIZE }, () => createConnection());
  const results = await Promise.allSettled(promises);

  const succeeded = results.filter((r) => r.status === 'fulfilled');
  const failed = results.filter((r) => r.status === 'rejected');
  const elapsed = Date.now() - startTime;

  console.log(`  Completed in ${elapsed}ms`);
  console.log(`  Succeeded: ${succeeded.length}/${STORM_SIZE}`);
  console.log(`  Failed: ${failed.length}/${STORM_SIZE}`);

  // Verify all successful connections got assigned to workers
  const connections = succeeded.map(
    (r) => (r as PromiseFulfilledResult<TestConnection>).value,
  );
  const orphaned = connections.filter((c) => !c.workerId);
  console.log(`  Orphaned (no worker): ${orphaned.length}`);

  if (orphaned.length > 0) {
    console.error('  FAIL: Some connections were not assigned to workers');
  }

  if (failed.length > 0) {
    console.error('  FAIL: Some connection creates failed');
  }

  // Cleanup
  for (const conn of connections) {
    await deleteConnection(conn.id).catch(() => {});
  }
}

// ---------------------------------------------------------------------------
// Test: Webhook Throughput
// ---------------------------------------------------------------------------

async function testWebhookThroughput(): Promise<void> {
  console.log('\n=== Test: Webhook Throughput ===');
  console.log('Simulating 1000 webhook events/second for 60 seconds...');

  const result = await fireWebhookEvents(1000, 60);

  console.log(`  Total sent: ${result.totalSent}`);
  console.log(`  Failures: ${result.failures}`);
  console.log(`  Avg latency: ${result.avgLatencyMs.toFixed(2)}ms`);

  const failureRate = result.failures / result.totalSent;
  if (failureRate > 0.01) {
    console.error(
      `  FAIL: Failure rate ${(failureRate * 100).toFixed(2)}% exceeds 1% threshold`,
    );
  }
}

// ---------------------------------------------------------------------------
// Test: Worker Failure Recovery
// ---------------------------------------------------------------------------

async function testWorkerFailureRecovery(): Promise<void> {
  console.log('\n=== Test: Worker Failure Recovery ===');
  console.log(
    'This test requires manual intervention or SSH access to kill a worker process.',
  );

  // Step 1: Create a connection, verify it is "working"
  const conn = await createConnection();
  console.log(`  Created connection ${conn.sessionName} on worker ${conn.workerId}`);

  // Step 2: Kill the worker process (manual step or via Hetzner API)
  console.log(
    `  ACTION REQUIRED: Kill worker ${conn.workerId} (e.g., sudo systemctl stop docker)`,
  );
  console.log('  Press Enter to continue after killing the worker...');
  // In automated form: await hetznerApi.resetServer(workerId);

  // Step 3: Wait for session to be detected as failed
  await waitFor(
    'session status to become "failed"',
    async () => {
      // GET /connections/:id and check status
      return false; // Placeholder
    },
    120_000,
  );

  // Step 4: Wait for auto-reconnect on a replacement worker
  await waitFor(
    'session to be reassigned and reconnected',
    async () => {
      // GET /connections/:id and check status === "working" and new workerId
      return false; // Placeholder
    },
    300_000,
  );

  console.log('  Worker failure recovery verified');

  // Cleanup
  await deleteConnection(conn.id).catch(() => {});
}

// ---------------------------------------------------------------------------
// Test: Session Persistence Across Restart
// ---------------------------------------------------------------------------

async function testSessionPersistence(): Promise<void> {
  console.log('\n=== Test: Session Persistence Across Restart ===');

  // Step 1: Create a connection and wait for it to be "working"
  const conn = await createConnection();
  console.log(`  Connection ${conn.sessionName} is working`);

  // Step 2: Restart the worker VM
  console.log(`  Restarting worker ${conn.workerId}...`);
  // await hetznerApi.rebootServer(conn.workerId);

  // Step 3: Wait for session to restore from Postgres
  await waitFor(
    'session to restore after worker restart',
    async () => {
      return false; // Placeholder — check session status returns to "working"
    },
    300_000,
  );

  console.log('  Session persistence verified');

  // Cleanup
  await deleteConnection(conn.id).catch(() => {});
}

// ---------------------------------------------------------------------------
// Test: Concurrent Modifications
// ---------------------------------------------------------------------------

async function testConcurrentModifications(): Promise<void> {
  console.log('\n=== Test: Concurrent Modifications ===');

  // Step 1: Create a single connection
  const conn = await createConnection();
  console.log(`  Testing concurrent modifications on ${conn.sessionName}`);

  // Step 2: Fire 20 simultaneous update requests
  const updates = Array.from({ length: 20 }, (_, i) =>
    fetch(`${CONFIG.apiUrl}/connections/${conn.id}`, {
      method: 'PATCH',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${CONFIG.testToken}`,
      },
      body: JSON.stringify({ name: `updated-${i}` }),
    }),
  );

  const results = await Promise.allSettled(updates);
  const succeeded = results.filter((r) => r.status === 'fulfilled');
  const failed = results.filter((r) => r.status === 'rejected');

  console.log(`  Succeeded: ${succeeded.length}/20`);
  console.log(`  Failed: ${failed.length}/20`);

  // Step 3: Verify final state is consistent (one of the updates won)
  // GET /connections/:id and verify no corruption

  // Cleanup
  await deleteConnection(conn.id).catch(() => {});
}

// ---------------------------------------------------------------------------
// Main Runner
// ---------------------------------------------------------------------------

async function main(): Promise<void> {
  console.log('Wago Scale & Stress Test Suite');
  console.log('=================================');
  console.log(`API URL: ${CONFIG.apiUrl}`);
  console.log(`Token: ${CONFIG.testToken ? '[provided]' : '[MISSING]'}`);
  console.log();

  if (!CONFIG.testToken) {
    console.error('ERROR: Set TEST_JWT_TOKEN environment variable');
    process.exit(1);
  }

  const tests = [
    { name: 'Scale-Up Trigger', fn: testScaleUpTrigger },
    { name: 'Scale-Down Trigger', fn: testScaleDownTrigger },
    { name: 'Connection Storm', fn: testConnectionStorm },
    { name: 'Webhook Throughput', fn: testWebhookThroughput },
    { name: 'Worker Failure Recovery', fn: testWorkerFailureRecovery },
    { name: 'Session Persistence', fn: testSessionPersistence },
    { name: 'Concurrent Modifications', fn: testConcurrentModifications },
  ];

  const selectedTest = process.argv[2];

  for (const test of tests) {
    if (selectedTest && !test.name.toLowerCase().includes(selectedTest.toLowerCase())) {
      continue;
    }

    try {
      await test.fn();
      console.log(`  PASS: ${test.name}`);
    } catch (err) {
      console.error(`  FAIL: ${test.name}`);
      console.error(`    ${err instanceof Error ? err.message : err}`);
    }
  }
}

main().catch(console.error);

File: tests/scale-test.py
"""
Wago Scale Test — End-to-End Cross-Worker Verification

Tests the full scale-up and scale-down lifecycle:
1. Baseline: verify clean state
2. Fill worker 1 (50 connections)
3. Trigger scale-up (create connection 51+)
4. Verify 2 workers active
5. Delete connections to trigger scale-down
6. Verify back to 1 worker
7. Verify Hetzner nodes cleaned up

Usage:
  WAGO_API_KEY=wh_... python tests/scale-test.py
"""

import asyncio
import os
import sys
import time

import httpx

API_KEY = os.environ.get("WAGO_API_KEY")
BASE_URL = os.environ.get("WAGO_API_URL", "https://api.wago.com")

if not API_KEY:
    print("Set WAGO_API_KEY environment variable")
    sys.exit(1)

passed = 0
failed = 0
errors = []


def ok(msg, detail=""):
    global passed
    passed += 1
    print(f"  ✓ {msg}" + (f"  ({detail})" if detail else ""))


def fail(msg, detail=""):
    global failed
    failed += 1
    errors.append(f"{msg}: {detail}")
    print(f"  ✗ {msg}: {detail}")


def info(msg):
    print(f"    {msg}")


async def create_connections(client, count, pace=0.5):
    ids = []
    fails = 0
    for i in range(count):
        try:
            r = await client.post("/connections")
            if r.status_code in (200, 201):
                ids.append(r.json()["id"])
            elif r.status_code == 429:
                await asyncio.sleep(2)
                r = await client.post("/connections")
                if r.status_code in (200, 201):
                    ids.append(r.json()["id"])
                else:
                    fails += 1
            else:
                fails += 1
        except Exception:
            fails += 1
        await asyncio.sleep(pace)
    return ids, fails


async def delete_connections(client, ids, pace=0.3):
    deleted = 0
    for cid in ids:
        try:
            r = await client.delete(f"/connections/{cid}")
            if r.status_code in (200, 201):
                deleted += 1
        except Exception:
            pass
        await asyncio.sleep(pace)
    return deleted


async def get_state(client):
    r = await client.get("/connections")
    conns = r.json() if r.status_code == 200 and isinstance(r.json(), list) else []
    active = [c for c in conns if c.get("status") != "stopped"]
    workers = {}
    for c in active:
        wid = c.get("workerId")
        if wid:
            workers[wid] = workers.get(wid, 0) + 1
    unassigned = sum(1 for c in active if not c.get("workerId"))
    return active, workers, unassigned


async def wait_for_condition(client, check_fn, description, timeout=180, interval=10):
    """Poll until check_fn returns True. Returns (success, state)."""
    info(f"Waiting for: {description} (timeout {timeout}s)...")
    start = time.monotonic()
    while time.monotonic() - start < timeout:
        active, workers, unassigned = await get_state(client)
        if check_fn(active, workers, unassigned):
            elapsed = time.monotonic() - start
            info(f"Condition met after {elapsed:.0f}s")
            return True, active, workers, unassigned
        await asyncio.sleep(interval)
    active, workers, unassigned = await get_state(client)
    return False, active, workers, unassigned


async def main():
    print("=" * 60)
    print("Wago Scale Test — End-to-End Cross-Worker")
    print(f"Target: {BASE_URL}")
    print("=" * 60)

    client = httpx.AsyncClient(
        base_url=f"{BASE_URL}/api",
        headers={"Authorization": f"Bearer {API_KEY}", "Content-Type": "application/json"},
        timeout=30.0,
    )

    all_ids = []

    # ── Phase 1: Baseline ──
    print("\n── Phase 1: Baseline ──")
    active, workers, _ = await get_state(client)
    if len(active) == 0:
        ok("clean baseline")
    else:
        info(f"⚠ {len(active)} pre-existing connections")

    # ── Phase 2: Fill worker 1 (50 connections) ──
    print("\n── Phase 2: Fill worker 1 with 50 connections ──")

    ids, fails = await create_connections(client, 50)
    all_ids.extend(ids)
    info(f"Created {len(ids)}/50 ({fails} failed)")

    success, active, workers, unassigned = await wait_for_condition(
        client,
        lambda a, w, u: sum(1 for c in a if c["id"] in all_ids and c.get("workerId")) >= 48,
        "≥48 connections assigned",
        timeout=120,
    )

    worker_ids = set(c.get("workerId") for c in active if c["id"] in all_ids and c.get("workerId"))
    assigned = sum(1 for c in active if c["id"] in all_ids and c.get("workerId"))
    info(f"Assigned: {assigned}/50, Workers: {len(worker_ids)}, Unassigned: {unassigned}")

    if assigned >= 48:
        ok(f"{assigned} connections assigned")
    else:
        fail("assignment", f"only {assigned}/50 assigned")

    if len(worker_ids) == 1:
        ok("all on 1 worker")
    else:
        fail("worker count", f"expected 1, got {len(worker_ids)}")

    # ── Phase 3: Trigger scale-up (create 10 more = 60 total) ──
    print("\n── Phase 3: Create 10 more to trigger scale-up (total 60) ──")

    ids2, fails2 = await create_connections(client, 10)
    all_ids.extend(ids2)
    info(f"Created {len(ids2)}/10 ({fails2} failed)")

    # Wait for second worker to appear (may take 2-3 min for Hetzner node)
    success, active, workers, unassigned = await wait_for_condition(
        client,
        lambda a, w, u: len(w) >= 2,
        "2 workers active",
        timeout=300,  # 5 min for Hetzner node provisioning
        interval=15,
    )

    worker_ids = set(c.get("workerId") for c in active if c["id"] in all_ids and c.get("workerId"))
    assigned = sum(1 for c in active if c["id"] in all_ids and c.get("workerId"))
    info(f"Workers: {len(worker_ids)}, Assigned: {assigned}/{len(all_ids)}")

    for wid, count in workers.items():
        info(f"  Worker {wid[:8]}...: {count} sessions")

    if len(worker_ids) >= 2:
        ok(f"scaled to {len(worker_ids)} workers")
    elif len(worker_ids) == 1 and unassigned > 0:
        # Worker 2 might not be ready yet — pending connections exist
        info(f"{unassigned} connections still unassigned (worker provisioning)")
        ok(f"scale-up initiated ({unassigned} pending assignment)", "worker 2 provisioning")
    else:
        fail("scale-up", f"only {len(worker_ids)} workers, {unassigned} unassigned")

    # ── Phase 4: Delete all connections ──
    print(f"\n── Phase 4: Delete all {len(all_ids)} connections ──")

    deleted = await delete_connections(client, all_ids)
    info(f"Deleted {deleted}/{len(all_ids)}")

    # Wait for cleanup
    success, active, workers, unassigned = await wait_for_condition(
        client,
        lambda a, w, u: sum(1 for c in a if c["id"] in all_ids) == 0,
        "all test connections gone",
        timeout=30,
    )

    remaining = sum(1 for c in active if c["id"] in all_ids)
    if remaining == 0:
        ok("all test connections deleted")
    else:
        fail("cleanup", f"{remaining} still active")

    # ── Phase 5: Wait for scale-down ──
    print("\n── Phase 5: Wait for scale-down (health poll + checkScaling) ──")

    # checkScaling runs every 5 minutes — wait up to 7 min
    success, active, workers, unassigned = await wait_for_condition(
        client,
        lambda a, w, u: len(w) <= 1,
        "≤1 workers with sessions",
        timeout=420,  # 7 minutes
        interval=30,
    )

    if len(workers) <= 1:
        ok(f"scaled down to {len(workers)} workers")
    else:
        fail("scale-down", f"still {len(workers)} workers after 7 min")

    # ── Phase 6: Final state ──
    print("\n── Phase 6: Final state ──")

    active, workers, unassigned = await get_state(client)
    info(f"Active connections: {len(active)}")
    info(f"Workers with sessions: {len(workers)}")

    test_remaining = sum(1 for c in active if c["id"] in all_ids)
    if test_remaining == 0 and len(workers) <= 1:
        ok("system returned to baseline")
    else:
        fail("final state", f"{test_remaining} test conns, {len(workers)} workers")

    await client.aclose()

    # ── Summary ──
    print("\n" + "=" * 60)
    total = passed + failed
    status = "PASS" if failed == 0 else "FAIL"
    print(f"{status}: {passed} passed, {failed} failed out of {total}")
    if errors:
        print("\nFailures:")
        for e in errors:
            print(f"  ✗ {e}")
    print("=" * 60 + "\n")
    sys.exit(1 if failed > 0 else 0)


if __name__ == "__main__":
    asyncio.run(main())
